File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int>tails;

        for(auto &x: nums){
            auto posItr = lower_bound(tails.begin(),tails.end(), x);

            if(posItr == tails.end())
                tails.push_back(x);
            
            else
                *posItr = x;
        }
        return tails.size();

    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
import bisect

class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        tails = []
        for x in nums:
            pos = bisect.bisect_left(tails, x)
            if(pos == len(tails)):
                tails.append(x)
            else:
                tails[pos] = x
        
        return len(tails)